### ISSUE TITLE
Basic iteration support / from_slice_* performance & safety.

### ISSUE THREAD

**Issue by ralfbiedert** [rust-lang-nursery/packed_simd#65]
Not sure how much of this should be handled through an RFC (or 3rd party libraries), but I feel basic iteration support could  increase performance, safety and ergonomics of `std::simd`: 

To iterate over a slice today, you have to call one of the `from_slice_*` methods. Either `from_slice_aligned`, which will do internal sanity checks per call, or `from_slice_aligned_unchecked`, which will just reinterpret the slice. 

Comparing them in a real-world application gives me these performance numbers:

```rust
fn compute_inner_kernel_simdf32x8(sv: &[f32], feature: &[f32], gamma: f32) -> f64 {
    type f32s = f32x8;

    let width = f32s::lanes();
    let steps = sv.len() / width;

    let mut sum = f32s::splat(0.0);

    for i in 0..steps {
        // When benchmarking `csvm_predict_sv1024_attr1024_problems1` with AVX2:

        // 238,928 ns / iter
        let a = unsafe { f32s::from_slice_aligned_unchecked(&sv[i * width..]) };
        let b = unsafe { f32s::from_slice_aligned_unchecked(&feature[i * width..]) };

        // 237,541 ns / iter
        // let a = unsafe { f32s::from_slice_unaligned_unchecked(&sv[i * width..]) };
        // let b = unsafe { f32s::from_slice_unaligned_unchecked(&feature[i * width..]) };

        // 343,970 ns / iter
        // let a = f32s::from_slice_aligned(&sv[i * width..]);
        // let b = f32s::from_slice_aligned(&feature[i * width..]);

        // 363,796 ns / iter
        // let a = f32s::from_slice_unaligned(&sv[i * width..]);
        // let b = f32s::from_slice_unaligned(&feature[i * width..]);

        // Add result
        sum += (a - b) * (a - b);
    }

    f64::from((-gamma * sum.sum()).exp())
}
  
```

In other words, using the safe checked version is at least 50% slower than the unsafe unchecked ones. 

I would prefer having a safe version, that is about as fast as the unchecked one. 

I think this could be achieved by having basic `[f32]` (and similar) iterators. Nothing as sophisticated as [faster](https://github.com/AdamNiederer/faster) that transparently handles partials, but maybe something like `f32x8::iterator(my_slice)` that would perform all sanity checks on the first call, and then (without further checks) just generates all subsequent elements.

**Comment by ralfbiedert** [comment]
> For a `&[f32]` slice, should we pick `f32x2`, `f32x4`, `f32x8`, `f32x16`? Should we pick one or the other depending on target features / target enabled at compile-time? Should we pick a different one depending on target features available at run-time? Should we generate code for the target features known at compile-time, or for different variants selected at run-time?

From my understanding of the previous discussions I see about 3 library levels here (please correct me if I'm wrong): 

- low-level `std::arch` / LLVM intrinsics
- mid-level `std::simd`,
- high-level `faster`

Those questions I'd probably expect to be handled in `faster`, or any other high level SIMD library; or an to-be-discussed extension of RFC2366. `std::simd` seems more about "safe SIMD for everything that's a multiple of N".

My point was more: 

`std::simd` provides a number of vectors (e.g, `f32x8`) with methods such as `from_slice_aligned`. These vectors have a fixed width and internally `assert` / panic if some preconditions loading that slice are not met (e.g., `slice.len() >= $elem_count`). 

I believe a common scenario is that you will have to `from_slice_aligned` multiple `f32x8` from a much bigger slice with a `len()` multiple of `8`  (all numbers / types just an example). 

In that case, you will either have to pay the `assert` cost multiple times (once per `f32x8` loaded), or resort to `unsafe`. 

If, instead, there were an iterator for each vector type (e.g., `f32x4::iterator()`, `f32x8::iterator()`, ... (names just for illustration)) that method `f32x8::iterator()` could the first time assert its preconditions (same as `f32x8::from_slice_aligned`  plus "check slice is multiple of `8`"), and then safely produce the rest of the elements without further checks. 

> What happens if the number of elements in the slice is not divisible by ::lanes() ? Should we not iterate over the tail? Should we give you a vector over the tail that has elements set to some default value like faster does? (how can your kernel tell if it is iterating over the tail?) Should we always give you a vector and a mask and require you to use select to operate only on the lanes for which the mask is true ? Etc.

I would love to have a library that "just works" with random slices, and I don't even mind `std::simd` becoming that library, but I guess that's a separate discussion from my original issue. 

> You can already build different types of these iterators on top of packed_simd, and that's what faster actually does (or used to do).

Yes. My only point is I would like to have "zero-cost iteration" of "slices multiple of N" without `unsafe`. 

> and @rkruppe has convinced me that none of the designs we choose here can actually be as good as an SPMD programming model for Rust (e.g. something like ISPC for Rust).

Just out of curiosity, is that discussion somewhere available?

**Comment by ralfbiedert** [comment]
Speaking of [faster](https://github.com/AdamNiederer/faster), there is actually a [tracking issue discussing how `std::simd` and `faster` should coexist](https://github.com/AdamNiederer/faster/issues/53).

**Comment by gnzlbg** [comment]
> I believe a common scenario is that you will have to from_slice_aligned multiple f32x8 from a much bigger slice with a len() multiple of 8 (all numbers / types just an example).
>
> In that case, you will either have to pay the assert cost multiple times (once per f32x8 loaded), or resort to unsafe.

If the only thing you want to do is create an iterator over a slice if all proper conditions are met, you can just transmute the `&[f32]` to a `&[f32x4]`/`&[f32x8]`/... :

```rust
fn try_as_f32x4(x: &mut [f32]) -> Option<&mut [f32x4]> {
    if x.len() % f32x4::lanes() != 0 || x.as_ptr().align_offset(mem::align_of::<f32x4>()) != 0 {
       None
    } else {
        // safe because a `f32x4` is layout compatible with `[f32; 4], 
        // each [f32; 4] within the [f32] slice is at mem::align_of::<f32x4>
        // and x.len() % f32x4::lanes() == 0
        unsafe { 
                Some(slice::from_raw_parts_mut(x.as_ptr_mut() as *mut f32x4, x.len() / f32x4::lanes()) 
        }
    }
}
```

Note that slices already implement `Iterator` (although you can call `.iter()`, etc.) so there is no need for us to implement any kind of `Iterator` here. 

You can probably make that generic over all vector types easily, maybe we could provide a way to do this in `std::simd` ?

However, if not all conditions are met (e.g. the slice is not properly aligned, the number of elements in the slice is not divisible by the vector length, etc.), then any `Iterator` over the slice is going to have to deal with all the issues (and probably more) discussed above.

> Just out of curiosity, is that discussion somewhere available?

It should be in the logs in #rust-internals or #rustc on IRC, but I don't recall when did it exactly happened (it was in the last two months) - I'm gnzlbg on IRC, so you might be able to find it by grepping the logs (I don't know where they are, but I they are hosted somewhere).

**Comment by AdamNiederer** [comment]
> I personally haven't found any of the designs I've seen convincing enough to say "this is good for most people or everybody", and @rkruppe has convinced me that none of the designs we choose here can actually be as good as an SPMD programming model for Rust (e.g. something like ISPC for Rust). The best we could do in a library is emulate it, and all proposals for it that I know have serious cons.

Agreed in full. It's a mess, and although I'm quite proud of what I was/am able to accomplish with faster, I wouldn't be comfortable putting any of the implementations I've come up with in some sort of officially-supported library, even if the implementation was polished to a mirror shine. Far too many footguns, wacky performance dips, codegen issues, uses of unsafe, etc. Something which works in "all cases" would have to be [horrifically overbuilt](https://github.com/EnterpriseQualityCoding/FizzBuzzEnterpriseEdition) and programmed to appease LLVM's optimizer (even in faster's current macro-laden state, I can't seem to get rid of two superfluous jumps in the hot parts of the code).

**Comment by ralfbiedert** [comment]
> If the only thing you want to do is create an iterator over a slice if all proper conditions are met, you can just transmute the &[f32] to a &[f32x4]/&[f32x8]/... 

Great, yes, that's all I'm asking for! 

Can we add these methods to `packed_simd` somewhere?

Again, I believe many users will face that problem, and I would hate to see everyone implementing the same unsafe code for it (or worse, forget an important check and run into UB like I did when toying with `from_slice_aligned_unchecked`). 

By the way, if the convention for `from_slice_aligned` is to panic, I'd be happy for these methods to panic as well, e.g.:

```rust
f32x8::f32s_from_slice_aligned(x: &mut [f32]) -> &mut [f32x8]
```

**Comment by gnzlbg** [comment]
Can you comment on the RFC PR with your proposal? I'd like to hear what others think about this before implementing anything since the feedback there tends to improve things a lot.

**Comment by ralfbiedert** [comment]
> Can you comment on the RFC PR with your proposal? I'd like to hear what others think about this before implementing anything since the feedback there tends to improve things a lot.

Will do!

**Comment by ralfbiedert** [comment]
Quick copy-paste as posted in the RFC thread:

> I now experimented a bit (see crate [**simd_aligned**](https://github.com/ralfbiedert/simd_aligned_rust)) and realized it is not as bad as I thought:
> 
> - Going from `&[f32x4]` to `&[f32]` is much nicer than trying the other way (in retrospect kind of obvious) and feels like "the right way" of doing it if I look at it from an application level. 
> - It only needs two relatively isolated `unsafe` calls to create a number of safe abstractions on top. 
> - These abstractions can easily be put in a thin 3rd party crate. 
> 
> While there could still be a minimal `packed_as_flat(data: &[Simd]) -> &[Simd::Element]` convenience function in `std::simd`, I now wouldn't mind if there is not.

In other words, I think my original request was not a good idea and this thread can be closed. Also, thank you very much for your feedback.
